### PR TITLE
mgmt/mcumgr: Commits that internalize cbor_nb_* functions and remove zephyr/mgmt/mcumgr/buf.h

### PR DESCRIPTION
**Pull request by de-nordic**
Three commits:
 1) internalizes cbor_nb_* functions as there is no reason to export them;
 2) Moves rest of the zephyr/mgmt/mcumgr/buf.h into subsys/mgmt/mcumgr/lib/smp/include/smp/smp.h and removes the former;
 3) Moves mcumgr_buf_alloc and mcumgr_buf_free out of buf.c to smp.c, renames them smp_packet_alloc and smp_packet_free, and remove the buf.c file completely.

### CODE REVIEW

File: subsys/mgmt/mcumgr/transport/smp_bt.c [hunk 1]
```diff
@@ -17,10 +17,11 @@
 #include <zephyr/bluetooth/gatt.h>
 
 #include <zephyr/mgmt/mcumgr/smp_bt.h>
-#include <zephyr/mgmt/mcumgr/buf.h>
 
 #include <zephyr/mgmt/mcumgr/smp.h>
+#include <zcbor_common.h>
```
Comment (nordicjm) [inline subsys/mgmt/mcumgr/transport/smp_bt.c:22]: No zcbor usage in this file

File: subsys/mgmt/mcumgr/smp_reassembly.c [hunk 2]
```diff
@@ -7,8 +7,8 @@
 #include <zephyr/kernel.h>
 #include <zephyr/sys/byteorder.h>
 #include <zephyr/net/buf.h>
-#include <zephyr/mgmt/mcumgr/buf.h>
 #include <zephyr/mgmt/mcumgr/smp.h>
+#include <zcbor_common.h>
```
Comment (nordicjm) [inline subsys/mgmt/mcumgr/smp_reassembly.c:11]: Not using zcbor functions in this file

File: subsys/mgmt/mcumgr/lib/mgmt/src/mgmt.c [hunk 3]
```diff
@@ -4,13 +4,14 @@
...
 
+#include <zephyr/sys/slist.h>
 #include <zephyr/sys/byteorder.h>
 #include <string.h>
 
-#include "mgmt/mgmt.h"
-#include <zephyr/mgmt/mcumgr/buf.h>
-#include <zcbor_encode.h>
 #include <zcbor_common.h>
+#include <zcbor_encode.h>
```
Comment (nordicjm) [inline subsys/mgmt/mcumgr/lib/mgmt/src/mgmt.c:12]: No zcbor functions being used in this file

File: subsys/mgmt/mcumgr/lib/smp/include/smp/smp.h [hunk 4]
```diff
@@ -25,14 +25,39 @@
 #ifndef H_SMP_
 #define H_SMP_
 
-#include "mgmt/mgmt.h"
```
Comment (nordicjm) [inline subsys/mgmt/mcumgr/lib/smp/include/smp/smp.h:28]: This needs a zcbor and zephyr kernel include

File: subsys/mgmt/mcumgr/smp.c [hunk 5]
```diff
@@ -8,11 +8,11 @@
 #include <zephyr/kernel.h>
 #include <zephyr/device.h>
 #include <zephyr/net/buf.h>
-#include <zephyr/mgmt/mcumgr/buf.h>
-#include "mgmt/mgmt.h"
-#include "smp/smp.h"
+#include <zcbor_common.h>
```
Comment (nordicjm) [inline subsys/mgmt/mcumgr/smp.c:11]: No zcbor usage in this file

File: subsys/mgmt/mcumgr/transport/serial_util.c [hunk 6]
```diff
@@ -13,8 +13,10 @@
 #include <zephyr/sys/byteorder.h>
 #include <zephyr/net/buf.h>
 #include <zephyr/sys/base64.h>
-#include <zephyr/mgmt/mcumgr/buf.h>
 #include <zephyr/mgmt/mcumgr/serial.h>
+#include <zcbor_common.h>
```
Comment (nordicjm) [inline subsys/mgmt/mcumgr/transport/serial_util.c:17]: No zcbor function usage in this file

File: subsys/mgmt/mcumgr/transport/smp_shell.c [hunk 7]
```diff
@@ -12,15 +12,16 @@
 #include <zephyr/kernel.h>
 #include <zephyr/init.h>
 #include <zephyr/net/buf.h>
-#include <zephyr/mgmt/mcumgr/buf.h>
+#include <zephyr/shell/shell.h>
+#include <zephyr/shell/shell_uart.h>
+#include <zephyr/drivers/uart.h>
+#include <syscalls/uart.h>
+#include <zcbor_common.h>
```
Comment (nordicjm) [inline subsys/mgmt/mcumgr/transport/smp_shell.c:19]: No zcbor usage in this file

File: subsys/mgmt/mcumgr/transport/smp_uart.c [hunk 8]
```diff
@@ -12,9 +12,10 @@
 #include <zephyr/kernel.h>
 #include <zephyr/init.h>
 #include <zephyr/net/buf.h>
-#include <zephyr/mgmt/mcumgr/buf.h>
 #include <zephyr/drivers/console/uart_mcumgr.h>
+#include <zcbor_common.h>
```
Comment (nordicjm) [inline subsys/mgmt/mcumgr/transport/smp_uart.c:16]: No zcbor usage in this file

File: subsys/mgmt/mcumgr/transport/smp_udp.c [hunk 9]
```diff
@@ -18,9 +18,10 @@
 #include <zephyr/net/socket.h>
 #endif
 #include <errno.h>
+#include <zcbor_common.h>
```
Comment (nordicjm) [inline subsys/mgmt/mcumgr/transport/smp_udp.c:21]: No zcbor usage in this file

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: subsys/mgmt/mcumgr/transport/smp_bt.c (hunk 1 above)
Comment (de-nordic) [reply subsys/mgmt/mcumgr/transport/smp_bt.c:22]: Hm. It really baffles me how come I did add the <zcbor_common.h> to so many files, where it is not needed.

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: subsys/mgmt/mcumgr/lib/smp/include/smp/smp.h (hunk 4 above)
Comment (de-nordic) [reply subsys/mgmt/mcumgr/lib/smp/include/smp/smp.h:28]: No it does not. Nothing in this header defines anything that is `struct net_buf` only `struct net_buf*`.
Definition of `struct net_buf` is only needed on implementation side.

Comment (nordicjm) [reply subsys/mgmt/mcumgr/lib/smp/include/smp/smp.h:28]: This places the onus of file includes on the person that includes this file, the end user should not have to add file includes to their files to be able to include this file without build errors

File: subsys/mgmt/mcumgr/lib/smp/include/smp/smp.h [hunk 10]
```diff
@@ -25,14 +25,41 @@
...
+ *
+ * @param nb                    The net_buf to free.
+ */
+void mcumgr_buf_free(struct net_buf *nb);
+
 struct smp_streamer;
-struct mgmt_hdr;
 
 /** @typedef smp_tx_rsp_fn
  * @brief Transmits an SMP response packet.
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply subsys/mgmt/mcumgr/lib/smp/include/smp/smp.h]: `mgmt_streamer ` no longer exists.

File: subsys/mgmt/mcumgr/lib/smp/include/smp/smp.h [hunk 11]
```diff
@@ -25,14 +25,42 @@
 #ifndef H_SMP_
 #define H_SMP_
 
-#include "mgmt/mgmt.h"
+#include <zephyr/net/buf.h>
+#include <zcbor_common.h>
```
Comment (nordicjm) [inline subsys/mgmt/mcumgr/lib/smp/include/smp/smp.h:29]: Needs `zephyr/mgmt/mcumgr/smp.h` include for zephyr_smp_transport

**Review by nordicjm — CHANGES_REQUESTED** [review]

**Review by nordicjm — DISMISSED** [review]

**Review by nordicjm — DISMISSED** [review]

**Review by carlescufi — DISMISSED** [review]

File: subsys/mgmt/mcumgr/smp.c (hunk 5 above)
Comment (nordicjm) [reply subsys/mgmt/mcumgr/smp.c:11]: Not sure why I marked this as resolved.

**Review by nordicjm — COMMENTED** [review]
`subsys/mgmt/mcumgr/smp.c` has a zcbor include with no zcbor usage in the file, it's been marked as resolved, but the include is still there

**Review by nordicjm — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**